File: sentimental-cash/cash.py
import cs50


def main():
    dollars = get_dollars()

    quarters = compute_quarters(dollars)
    print(f"quarters = {quarters}")
    dollars = round(dollars - quarters * 0.25, 2)

    dimes = compute_dimes(dollars)
    print(f"dimes = {dimes}")
    dollars = round(dollars - dimes * 0.10, 2)

    nickels = compute_nickels(dollars)
    print(f"nickels = {nickels}")
    dollars = round(dollars - nickels * 0.05, 2)

    pennies = compute_pennies(dollars)
    print(f"pennies = {pennies}")
    dollars = round(dollars - pennies * 0.01, 2)

    # Sum coins
    coins = quarters + dimes + nickels + pennies

    # Print total number of coins to give the customer
    print(f"{coins}")


def get_dollars():
    """Ask how many dollars the customer is owed"""
    while (True):
        dollars = cs50.get_float("Change owned: ")
        if dollars > 0:
            break
    return dollars


def compute_coins(coins, coin_value):
    coin_value = int(100 * coin_value)
    coins = int(100 * coins)
    coins_count = 0
    for _ in range(coin_value, coins + 1, coin_value):
        coins_count += 1
    return coins_count


def compute_quarters(dollars):
    """Calculate the number of quarters to give the customer"""
    return compute_coins(dollars, 0.25)


def compute_dimes(dollars):
    """Calculate the number of dimes to give the customer"""
    return compute_coins(dollars, 0.10)


def compute_nickels(dollars):
    """Calculate the number of nickels to give the customer"""
    return compute_coins(dollars, 0.05)


def compute_pennies(dollars):
    """Calculate the number of pennies to give the customer"""
    return compute_coins(dollars, 0.01)


if __name__ == "__main__":
    main()

File: cash/cash.c
#include <cs50.h>
#include <stdio.h>

int get_cents(void);
int calculate_quarters(int cents);
int calculate_dimes(int cents);
int calculate_nickels(int cents);
int calculate_pennies(int cents);

int main(void)
{
    // Ask how many cents the customer is owed
    int cents = get_cents();

    // Calculate the number of quarters to give the customer
    int quarters = calculate_quarters(cents);
    cents = cents - quarters * 25;

    // Calculate the number of dimes to give the customer
    int dimes = calculate_dimes(cents);
    cents = cents - dimes * 10;

    // Calculate the number of nickels to give the customer
    int nickels = calculate_nickels(cents);
    cents = cents - nickels * 5;

    // Calculate the number of pennies to give the customer
    int pennies = calculate_pennies(cents);
    cents = cents - pennies * 1;

    // Sum coins
    int coins = quarters + dimes + nickels + pennies;

    // Print total number of coins to give the customer
    printf("%i\n", coins);
}

int get_cents(void)
{
    int cents;
    do
    {
        cents = get_int("Cents: ");
    }
    while (cents < 0);
    return cents;
}

int calculate_coins(int coins, int coin_value)
{
    int coins_count = 0;
    for (int i = coin_value; i <= coins; i += coin_value)
    {
        coins_count++;
    }
    return coins_count;
}

int calculate_quarters(int cents)
{
    return calculate_coins(cents, 25);
}

int calculate_dimes(int cents)
{
    return calculate_coins(cents, 10);
}

int calculate_nickels(int cents)
{
    return calculate_coins(cents, 5);
}

int calculate_pennies(int cents)
{
    return calculate_coins(cents, 1);
}
